### DISCUSSION TITLE
`Record`/`Field` Factories

### DISCUSSION THREAD

**Discussion by harrysolovay** [paritytech/scale-ts#11] [API Design]
Why is it that we define the shape of a record codec as follows?

```ts
const c = s.record(
  s.field("name", s.str),
  s.field("nickName", s.str),
);
```

At first glance, it seems strange that we must call `s.field`... couldn't we bypass this bloat?

Why wouldn't the following work?

```ts
const c = s.record({
  name: s.str,
  nickName: s.str,
});
```

Unfortunately, the field order needs to be known for encoding. There's no way for us to reflect on this info **[EDIT] well** using a JS record. We could certainly allow fields to be defined via tuples:

```ts
const c = s.record(
  ["name", s.str],
  ["nickName", s.str],
  // ...
);
```

The issue with this is that string literals "name" and "nickName" are widened to `string`.

```ts
type C = s.Native<typeof c>;

assertTypesEqual<C, {
  name: string;
  nickName: string;
}>();

// ^ this would produce an error, as the actual signature is `Record<string, string>`
```

We must assert `as const` in order to get the literal typings of keys, which flow into the signature of `c`.

```ts
const c = s.record(
  ["name", s.str] as const,
  ["nickName", s.str] as const,
  // ...
);
```

It's very easy to forget to `as const`-ify––better to safeguard via the extra factory `field`.

Additionally, the `Field` type / `field` factory is used elsewhere (tagged unions).

**Comment by josepot** [comment]
> Unfortunately, the field order needs to be known for encoding. There's no way for us to reflect on this info using a JS record

Why not?

JS [**does** have a guaranteed property order](https://tc39.es/ecma262/#sec-ordinaryownpropertykeys). This is something that was standardized on ES2015 and even before then it was a de facto standard across JS engines. That's why `@unstoppablejs/scale` admits JS Objects on its `Struct` higher order codec.

**Reply by harrysolovay** [reply]
This is a great comment & makes sense: seems ES2020 really solidified order access for object literals... but there are other ways in which devs may initialize codecs. Just to be safe, I'd prefer explicit indication of order... what do others think?

**Reply by harrysolovay** [reply]
[This playground](https://tsplay.dev/NBk4Vm) showcases a shortcoming of implicit ordering: the field order of objects of multiple key types does not seem to be preserved. It seems they are grouped. We want to enable decoding to / encoding from any `PropertyKey`. Thoughts @josepot?

**Reply by harrysolovay** [reply]
Another issue: many devs configure key sorting via formatters ([for instance](https://eslint.org/docs/rules/sort-keys)). They would need to disable such formatting.

**Comment by harrysolovay** [comment]
Addressed in PR #17 – locking conversation
